### ISSUE TITLE
scopedNamespace: Is this limited to a single namespace or a comma delimited list?

### ISSUE THREAD

**Issue by tommyreilly** [external-secrets/external-secrets#523]
**Describe the solution you'd like**
Describe the end goal of this proposal. What is this new functionality or the new behaviour (or what problem does it fix)? 

Our company are looking to use External Secrets in a GKE cluster that would be used by many of our product teams. Essentially, we will be hosting an External Secrets instance per product team in the GKE cluster, where each instance will use a product team specific workload identity to point at their own Secret Manager service in their own GCP service project (i.e. each product team have their own GCP service project).

Each product team will be deploying their workloads to more than one namespace in the GKE cluster. Thus, we want each instance of External Secrets to "watch" the namespaces used by a product team (where there will usually be more than one).

Does scopedNamespace allow the definition of more than one namespace? If not, why is it restricted to a single namespace and is it practical to request a feature enhancement to support multiple namespaces. We are switching from the use of KES, as advised by the community, but this multi-tenancy limitation of just watching a single namespace was not a restriction when using KES.

**What is the added value?**
Explain the value that it adds. e.g. "Secret refreshing will make internal secrets up to date with external secrets".

Improves the multi-tenancy implementation in a single k8s cluster.

**Give us examples of the outcome**

One k8s cluster
Team A and Team B deploy their workloads to the k8s cluster and both want to use External Secrets
Operations team deploy an instance of External Secrets for Team A and a separate instance for Team B to the k8s cluster
Team A: Uses namespace1 and namespace 2
Team B: Uses namespace3 and namespace4
Operations team want to list "namespace1,namespace2" for the scopedNamespace field in values.yaml for the Team A External Secrets instance and "namespace3,namespace4" for the scopedNamespace field in values.yaml fo the Team B External Secrets instance.

Provide templates if you are proposing changes in the CRD. Provide example workflows or code snippets if they make sense to present.

**Observations (Constraints, Context, etc):**

Give here all extra information that could be interesting. Such as Golang version and Kubernetes version if you are reporting a bug/problem. You can also foresee technical constrains like "this could only be implementing using this specific technology or approach, because of this and that".

**Comment by knelasevero** [comment]
Hey folks, thanks for your interest in the project!

I think you are confusing how to configure the project. Most of the "final" functionality of the project will be configured by creating the Custom Resources in the cluster, and not the configs of the helm chart. 

In KES you only had one CRD, named ExternalSecret (externalsecrets.kubernetes-client.io), in ESO we have ExternalSecret (externalsecrets.external-secrets.io), SecretStore (secretstores.external-secrets.io) and ClusterSecretStore (clustersecretstores.external-secrets.io).

For multi-tenancy, have a look over this documentation: https://external-secrets.io/guides-multi-tenancy/

You probably want to use one SecretStore per namespace.

We also started to ship a tool to make it easier to migrate from KES: https://github.com/external-secrets/kes-to-eso

But I guess it won't be that useful in your case specifically. :thinking: 

cc @gusfcarvalho

**Comment by gusfcarvalho** [comment]
I think it is worth to point some differences between KES and ESO:

KES can only have one ''provider account'' attached to a controller, during deploy time. So, in order to have multiple workload identity configuration, multiple KES deployments are needed.

ESO can have "multiple accounts" attached to the same controller, at runtime. This is done through configuration of a secretStore to use different service accounts (which then are annotated with different permissions).


TL;DR:
Option 1) (my personal recommendation)

- Create `ClusterSecretStore` A bound to `Workload Identity` of Team A
- Create Kyverno/Gatekeeper rules to enforce Team A can only use `ClusterSecretStore` A
- Repeat the process for any other teams or needs (and use ESO deployment as is)

Option 2) (closer to KES setup)
- Create custom RBAC for ESO controller `team-a` so it can only access specific Service Accounts. Deploy it using controller label `team-a`
- Create `ClusterSecretStore` A bound to Workload Identity of Team A, configured to use ESO controller `team-a`
This will enforce that only ESO Controller `team-a` could possibly access information from Team A Workload Identity.
- Repeat the process for any other Teams.

Longer Version:
For your specific use case, I agree with @knelasevero. The "ESO way" of having this setup as close as possible to your intent would be to create a SecretStore bound to a ServiceAccount with WorkloadIdentity permissions to namespace1 and namespace2, and then have team a using these SecretStores. SecretStores are currently namespaced, so you would need to have one per namespace (which I think is the main issue you're having here, as teams can have potentially hundreds of namespaces). Alternatively, you can create a ClusterSecretStore bound to a serviceAccount with specific WorkloadIdentity permissions. This ClusterSecretStore would impersonate access to Team's A, but could be used by Team B if no policy is implemented. My personal recommendation would be to use something like `Kyverno` to enforce these policies, but you can achieve that with External Secrets as well by configuring the Cluster Secret Store to a given ESO controller, and then have multiple ESO deployments (each with its own controller label). You'd still need to take into account appropriate RBAC permissions to avoid impersonations from unwanted SAs.

Regarding kes2eso, it can actually help you out with manifest configuration, as by default it creates all resources based on KES service accounts. You can pass `--kes-deployment-name` and `--kes-namespace` to force the tool to look for a particular KES deployment, and it will create a `ClusterSecretStore` and `ExternalSecrets` that are configured to use that specific KES ServiceAccount. The manual change you'll need to do is to point to a new SA (with the same WI annotations), and specify a controller for the ClusterSecretStores that are generated by KES (and be sure to configure the right match between a group of KES-ExternalSecrets and Its appropriate KES Deployment).

**Comment by tommyreilly** [comment]
Thank you @knelasevero @racdev @gusfcarvalho 

I get the general gist of the advice but still a little confused.

I'm good with using a SecretStore per namespace and would not want to go with the ClusterSecretStore due to the RBAC considerations you refer to - defining admissions controller policy on the k8s with Kyverno or something similar is not something we have in scope to implement (yet), so I'd like to limit this to namespace logic.

The challenge I see with the current implementation and the use case I describe is that the workload identity is "bound" to the eso controller - e.g. if it is deployed to a namespace named es, then that is where the k8s serviceaccount is used for the workload identity mapping to a Google Service Account (GSA). Ideally, if the authentication was to be managed within the namespace where the SecretStore is being defined such that a local k8s sa mapped to a GSA, then the one eso controller could be used.

The multi-tenancy documentation at https://external-secrets.io/guides-multi-tenancy/ has the ESO Service section with a diagram that visually suggests that SecretStores in different namespaces can access different external stores (diagram shows AWS but logically no different than accessing Secret Manager in GCP). The diagram, I assume, is correct if the authentication defined in each SecretStore defined, say, a k8s secret with the specific credentials in a k8s serviceaccount (as described at the top of the page at https://external-secrets.io/provider-google-secrets-manager/ (Service account key authentication).

But when using a workload identity, the SecretStore definition doesn't any content other than to defer to the serviceaccount that the eso controller runs as:

apiVersion: external-secrets.io/v1alpha1
kind: SecretStore
metadata:
  name: example
spec:
  provider:
    gcpsm:
      projectID: pid
      
Ideally, in the SecretStore, the workload identity k8s sa could be referenced with it expecting it to be local to the namespace and used for authentication (so that it would be at the same logical level if defining a sa with a json key to authenticate the SecretStore).

I don't want product teams to add a JSON key into a k8s secret in their namespaces (for security considerations of downloading a credential from GCP) and much prefer for them to define and use a workload identity for authentication of a SecretStore in the namespace(s) they are deploying to.

Doing this with a single eso instance in the GKE cluster rather than doing one per product team would be ideal.

Hoping I'm making sense :)

Thanks

Tommy

**Comment by knelasevero** [comment]
You make a lot of sense.

I guess you want to +1 this issue request ;) 
https://github.com/external-secrets/external-secrets/issues/372

Would love to get contributions there as well!

**Comment by tommyreilly** [comment]
@knelasevero - yes definitely a +1 from me.

My go skills are a bit scratchy, as otherwise I'd have a stab at this myself. If someone else has the inclination to look at this, that would be ideal.

**Comment by gusfcarvalho** [comment]
@tommyreilly  I think a workaround for now can be 'a little bit of both'. I think you could set ESO controller labels per team / workload identity and create namespaced secretstores pointing to those controller labels. You'd still need 1 ESO controller per team, but you wouldn't have to face RBAC issues anymore (because the secretstore now exists only in a specific namespace).

but yeah, ideally its #372 =)

**Comment by racdev** [comment]
So my use case is that we only use HashiCorp Vault and I have a "platform team" that manages services in namespaces a, b and c, each with a service account a, b and c in each namespace respectively. Each SA is bound to a role in Vault with relevant policies to access the secrets needed for the services running in the respective namespace. 

I therefore deploy the ESO controller in namespace es running as SA es that has the relevant RBAC to read the SA's in namespaces a, b and c - I can then create SecretStores in namespaces a, b and c referring to the relevant namespace SA and role and then create my namespaced secrets. All good.

My understanding however is, because the controller is cluster scoped in terms of namespaces it watches, team x using namespace x could create a SecretStore in namespace x and refer to the SA's in namespaces a, b and c, and because the controller sees all SecretStores in all namespaces and has the relevant RBAC to access the SAs in namespaces a, b and c, the SecretStore would work and team x has access to the "platform teams" secrets. Not good.

If the controller was namespace scoped to only refer to namespaces a, b and c, then it would only ever "see" the platform teams SecretStores and hence no other team could ever create a SecretStore that uses the SAs in the platform teams namespaces as those SecretStores would be invisible to the controller running in the es namespace. Each team would obviously have to have their own controller - again scoped to just their namespaces - but this would provide a common pattern of one controller per team, scoped to the teams namespaces with a SA with access to the SAs in the managed namespaces. This would remove any issue of cross namespace access, without the need for another service to police usage. - whilst letting teams do what they like SecretStore and ExternalSecret wise.

Apologies if my understanding is not correct.

**Comment by moolen** [comment]
> [...] team x using namespace x could create a SecretStore in namespace x and refer to the SA's in namespaces a, b and c [...]
The controller reconciles the `ExternalSecret` and `SecretStores` resources in all namespaces. However, A SecretStore can only use SAs from it's own Namespace. SAs from other namespaces can not be used. This is by design: Namespaced resources (and references) should not transcend namespaces. 

Scoping the controller to a list of namespaces is desirable to limit the reconciling of resources to a subset of namespaces (and hence eliminate the issue of a potential software bug that allows users to use SAs across namespaces).

**Comment by knelasevero** [comment]
That's true. Even though the discussion evolved to the solution that delivers what users are looking for, scoping some namespaces is still something that needs to happen. Do we have some other issue for it?

**Comment by racdev** [comment]
> However, A SecretStore can only use SAs from it's own Namespace. SAs from other namespaces can not be used

Ah, that clarifies things. The bit that confused me was that the SecretStore configuration has namespace as a field when referring to the Service Account which suggests it can refer to a SA in another namespace than the namespace of the SecretStore itself:
```
        # Optional service account reference
        serviceAccountRef:
          name: "my-sa"
          namespace: "secret-admin"
```

**Comment by moolen** [comment]
That's a documentation flaw because the underlying type is shared with the `ClusterSecretStore` where you can specify the field (it's ignored in a `SecretStore`).

**Comment by tommyreilly** [comment]
@gusfcarvalho - thank you, at this time, I'll try and see if I can get your suggested workaround working for our use case.

**Comment by gusfcarvalho** [comment]
Closing this one so we can track the remaining bit of it at #163!
